conf/schemas.go

```diff
@@ -58,50 +91,114 @@ func ReadSchemas(file string) (Schemas, error) {
 		}
 
 		if sec.ValueOf("pattern") == "" {
-			return nil, fmt.Errorf("[%s]: empty pattern", schema.Name)
+			return Schemas{}, fmt.Errorf("[%s]: empty pattern", schema.Name)
 		}
 		schema.Pattern, err = regexp.Compile(sec.ValueOf("pattern"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
 		}
 
 		schema.Retentions, err = ParseRetentions(sec.ValueOf("retentions"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
 		}
 
 		p := int64(0)
 		if sec.ValueOf("priority") != "" {
 			p, err = strconv.ParseInt(sec.ValueOf("priority"), 10, 0)
 			if err != nil {
-				return nil, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
+				return Schemas{}, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
 			}
 		}
 		schema.Priority = int64(p)<<32 - int64(i) // to sort records with same priority by position in file
 
 		schemas = append(schemas, schema)
 	}
 
-	sort.Sort(schemas)
-	return schemas, nil
+	sort.Sort(SchemaSlice(schemas))
+
+	return NewSchemas(schemas), nil
 }
 
 // Match returns the correct schema setting for the given metric
 // it can always find a valid setting, because there's a default catch all
 // also returns the index of the setting, to efficiently reference it
-func (s Schemas) Match(metric string) (uint16, Schema) {
-	for i, schema := range s {
+func (s Schemas) Match(metric string, interval int) (uint16, Schema) {
```

Comment:
I'd like the suggestion B more mainly for performance reasons. If i understand it right in the current solution let's say if we have `3` patterns and each has `3` retentions, then if something calls `Match()` with a metric/interval that matches the last retention of the last pattern, then `.MatchString()` will be called `9` times because it's called for each retention of each pattern. Assuming that pattern matching is relatively expensive compared to an index lookup it would seem more efficient to have a nested structure that allows us to only call `.MatchString()` once per pattern.